pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationDataSource.java

```diff
@@ -82,39 +83,77 @@ default String getHttpHeader(String name) {
 
     /**
      * Check if data from Pulsar protocol are available.
-     * 
+     *
      * @return true if this authentication data contain data from Pulsar protocol
      */
     default boolean hasDataFromCommand() {
         return false;
     }
 
     /**
-     * 
+     *
      * @return authentication data which is stored in a command
      */
     default String getCommandData() {
         return null;
     }
 
+
+    /*
+     * SASL
+     */
+
+    /**
+     *
+     * @return authentication data which will be stored in a command
+     */
+    default byte[] getCommandDataBytes() {
```

Comment:
shall we just standardize to use `byte[]`?

Reply:
We should do this. since String convert meet some error in this pr

Reply:
In other auth plugins we're base64 encoding the credential. That would not need any change in the plugin interfaces

Reply:
Thanks. It is Ok for me. If byte[] will involve too many changes, will try and use base64 later.

Reply:
@merlimat, one concern here is: Is it worth to convert between String and Byte using base64?
As in PulsarApi.proto, we store auth data as bytes in [CommandConnected](https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/proto/PulsarApi.proto#L197), and also in [SaslServer](https://docs.oracle.com/javase/7/docs/api/javax/security/sasl/SaslServer.html) and [SaslClient](https://docs.oracle.com/javase/7/docs/api/javax/security/sasl/SaslClient.html), it do the valuate and challenge also use bytes like this:
```
byte[]  evaluateResponse(byte[] response)
```
This means, we need to convert `bytes(proto cmd) <-> String(pulsar api) <-> bytes(sasl api)` each time. 
The benefit is the alignment with old api, while the opposite is useless converting.